aea/runtime.py

```diff
@@ -261,165 +247,79 @@ def set_loop(self, loop: AbstractEventLoop) -> None:
 
         :param loop: event loop to use.
         """
-        super().set_loop(loop)
-        self.multiplexer.set_loop(self.loop)
-        self.main_loop.set_loop(self.loop)
-        self._async_stop_lock = asyncio.Lock()
-
-    def _start(self) -> None:
-        """
-        Start runtime synchronously.
+        BaseRuntime.set_loop(self, loop)
 
-        Set event loops for multiplexer and agent run loop.
+    async def run(self) -> None:
+        """Start multiplexeor task."""
+        terminal_state = RuntimeStates.error
+        try:
+            await self.run_runtime()
+        except StopRuntime as e:
+            self._state.set(RuntimeStates.stopping)
+            terminal_state = RuntimeStates.stopped
+            if e.reraise:
+                raise e.reraise
+        except (asyncio.CancelledError, CancelledError, KeyboardInterrupt):
+            self._state.set(RuntimeStates.stopping)
+            terminal_state = RuntimeStates.stopped
+        finally:
+            await self.stop_runtime()
+            self._state.set(terminal_state)
 
-        Start runtime asynchonously in own event loop.
+    async def stop_runtime(self) -> None:
         """
-        self.set_loop(self.loop)
+        Stop runtime coroutine.
 
-        logger.debug(f"Start runtime event loop {self.loop}: {id(self.loop)}")
-        self._task = self.loop.create_task(self.run_runtime())
+        Stop main loop.
+        Agent teardown.
+        Disconnect multiplexer.
+        """
+        self.main_loop.stop()
+        with suppress(StopRuntime):
+            await self.main_loop.wait_completed()
+        self._teardown()
 
-        try:
-            self.loop.run_until_complete(self._task)
-            logger.debug("Runtime loop stopped!")
-        except Exception:
-            logger.exception("Exception raised during runtime processing")
-            self._state.set(RuntimeStates.error)
-            raise
-        finally:
-            self._stopping_task = None
+        self.multiplexer.stop()
+        await self.multiplexer.wait_completed()
```

Comment:
I suppose this will eventually collapse in a single call? E.g. `stop()` would be `async`

Reply:
The problem is that currently `self.multiplexer` is a `Multiplexer`, and not a `AsyncMultiplexer`. Related to David's comment: https://github.com/fetchai/agents-aea/pull/1731/files#r487761544

Another comment: should this have `with suppress(StopRuntime)` too?

Reply:
no, it's AsyncMultiplexer now